### PR TITLE
fsdp_copy_out proof of concept

### PR DESCRIPTION
**Pull request by yifuwang**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #115515

~Benchmark script: P901599447~
~\~500GB/s throughput for fsdp copy_out, straight from all_gather output to individual parameters, considering all overhead.~

~The approach requires padding individual parameters to align with 128 bits. Storage-wise, this comes at no extra cost because the minimum allocation size of CUDACachingAllocator is 512 bytes. Transportation-wise, the upper-bound increase in send size for bf16 is NUM_PARAMS * 7 elements. To put this in perspective, consider the benchmark script that test against a 0.15GB all_gather output; in this case, each device would only need to send an additional 302 bytes.~

No alignment requirement anymore. Can achieve 650-700GB/s on the tested sizes.

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @fegin @XilunWu @wanchaol @fduwjj @wz337 @tianyu-l @wconstab @yf225

### CODE REVIEW

**Comment by awgu** [comment]
Not important but maybe interesting: On a super small-scale case (3-layer transformer decoder on 2 GPUs, <1 GB model size), `fsdp_all_gather_copy_out_no_align_2` is slower than the naive torch implementation using for loop + `torch.cat`:

<details>
<summary> fsdp_all_gather_copy_out_no_align_2: 0.45 ms </summary>

Screenshot 2023-12-13 at 11 27 13 AM
</details>

<details>
<summary> for loop + torch.cat: 0.26 ms </summary>

Screenshot 2023-12-13 at 11 28 03 AM
</details>

**Comment by yifuwang** [comment]
Landed as an aten op fast path in https://github.com/pytorch/pytorch/pull/117203
